## Task 1.1 Summarize the GWAS results


In this first part of the assignement, we are going to load the summary results from a Genome-Wide Association Study (GWAS) and examine its results. In the second part, we examine the mouse homologous gene.

In [235]:
### THE TEXT AND CODE IN THIS FIELD IS PROVIDED TO THE STUDENTS
#
### Read ALL instructions FIRST before you begin your data analysis
#
# We are going to examine the GWAS ouput (i.e. summary statistics) of the 2018 GWAS of Major Depression conducted by the PGC.
# The data belong to this publication: Nat Genet. 2018 May;50(5):668-681. doi: 10.1038/s41588-018-0090-3. PMID:29700475
#
# Download this summary statistics file:
! wget "https://www.dropbox.com/scl/fi/yj3m8d2puymhdzufanhs9/MDD2018_ex23andMe.txt?rlkey=4dl6gdi977dyj6neuwjtdf1xv&st=82vc8u0o&dl=0"  -O MDD_2018_original.tsv

# Download the README file:
! wget "https://www.dropbox.com/scl/fi/0buvztu2sl6xa12vwvwyb/pgc-mdd-2018-readme.pdf?rlkey=r6qdropq18mi5rowplhoagcp8&st=v4i71xip&dl=0" -O MDD_2018_README.pdf

# Read the README file before moving to Task 1.1



--2025-12-11 16:28:04--  https://www.dropbox.com/scl/fi/yj3m8d2puymhdzufanhs9/MDD2018_ex23andMe.txt?rlkey=4dl6gdi977dyj6neuwjtdf1xv&st=82vc8u0o&dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.3.18, 2620:100:601b:18::a27d:812
Connecting to www.dropbox.com (www.dropbox.com)|162.125.3.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://uc1597a280f2a69cb34a813f8031.dl.dropboxusercontent.com/cd/0/inline/C21Cn8geOvHU4IjTVn8DA2BjLZLR0E5vY-SeIVm_PGZBvsMrvHPM9sH6G9GWNC1rNsWKUX2NgccLC0i45ykhEfoABTCBOcYVEkpvzR4Gp2m59kFk3kBYfWLoTC28oOqhRbY/file# [following]
--2025-12-11 16:28:04--  https://uc1597a280f2a69cb34a813f8031.dl.dropboxusercontent.com/cd/0/inline/C21Cn8geOvHU4IjTVn8DA2BjLZLR0E5vY-SeIVm_PGZBvsMrvHPM9sH6G9GWNC1rNsWKUX2NgccLC0i45ykhEfoABTCBOcYVEkpvzR4Gp2m59kFk3kBYfWLoTC28oOqhRbY/file
Resolving uc1597a280f2a69cb34a813f8031.dl.dropboxusercontent.com (uc1597a280f2a69cb34a813f8031.dl.dropboxusercontent.com)... 162.125.3.15, 2620:100:6023:15::a27d

**Q1.1.** Load the **tab** separated file 'MDD_2025_gwas_data.tsv' into a table and **check the content and the format**. Show your code and explain what you are doing and why. [2 points]

Relevant column headers:
 1 CHROM: Chromosome;
 2	POS: Base-pair position (GRCh37);
 3	SNP: SNP Marker ID (rs identifier);
 8	PVAL: P-value;

**Q1.2.** How many SNPs were tested?  [1 point]

**Q1.3.** Find the **most significant SNP**. What is this top SNP's rs identifier? [1 point]

**Q1.4.** Using the NCBI online database, in which gene is SNP rs9479138 located? [0.5 point]

**Q1.5.** Looking further in the NCBI database, in which body tissue or organ is this gene most expressed? Provide a link/url showing how you got to this answer. [0.5 point]

In [ ]:
import pandas as pd
df = pd.read_csv('MDD_2018_original.tsv', sep='\t')

/tmp/ipython-input-1174509403.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('MDD_2018_original.tsv', sep='\t')


In [ ]:
df.info()
print(df.head())
df.shape

Q1.1 I loaded the file with pandas and used the function df = pd.read_csv('MDD_2018_original.tsv', sep='\t'). With sep='\t') I generate seperately tabs for the columns instead with commas. With df.info() and df.head() I get the first 5 rows and overall information about the columns.

In [ ]:
print(f"Total SNPs tested: {df['SNP'].nunique()}")

Q1.2 Total number of unique SNPs tested is 13554489

In [ ]:
snp = df.loc[df['P'].idxmin()]

most_significant_SNP = snp['SNP']
most_significant_SNP_p_value = snp['P']

print(f"Most significant SNP: {most_significant_SNP}")
print(f"p-value: {most_significant_SNP_p_value}")



Q1.3 The most signficant SNP is rs76025409 with a p-value of 2.135e-11.

Q1.4 SNP rs9479138 is located in the ESR1 (Estrogen Receptor 1) gene region at Chr6

Q1.5 The ESR1 gene is most expressed based on the RNA of normal human tissues in the uterus and endometrium (HPA) https://www.ncbi.nlm.nih.gov/gene/2099#gene-expression


## Task 1.2 Microarray database
Another gene that has often been associated with Major Depression is **PGBD1**.In the Allen Institute microarray database, search for expression patterns of PGBD1: https://human.brain-map.org/microarray/search.

Download the relevant data for PGB1 (first probe, first row). There are two ways to do this:
1. Select the top row (probe) and download this data. Upload the data into  Colab by clicking on the 'file' icon on the top left corner.
2. Run the code provided below using wget commands (pointing to a dropbox database).

In [ ]:
#### This is provided to the students! #######
#downloading data
! wget 'https://www.dropbox.com/scl/fi/th8ykvjnk9rlao0ccp47d/Columns.csv?rlkey=dttm23lamkhugkypb0r5bgaj2&st=1wkbgc7r&dl=0' -O Columns.csv
! wget 'https://www.dropbox.com/scl/fi/xs9g0r3egrxi93hhn9pel/Expression.csv?rlkey=jf47vz3ez0mxeascgjdqjn6rc&st=92e3dfew&dl=0' -O Expression.csv
! wget 'https://www.dropbox.com/scl/fi/j2cn98z1vu5igppx69ayf/Contents.txt?rlkey=ew526p39wpe78os3x46rzglrg&st=wd6jzr40&dl=0' -O Contents.txt

**Read the "Contents.txt" file.**

**Load the Expression.csv file and the Columns.csv file in python.**

**Q1.6**  Format the data of Expression.csv and Columns.csv so that you have a table with 1 column of PGB1 expression values, next to columns with information of regions and subjects (donors). You should have 1 row per brain structure for each subject. (1 point)

**Q1.7** Calculate the mean expression accross participants per brain structure.  In which brain structure does this PGB1 have the highest mean expression? (0.5 point)

**Q1.8**   How many missing values are there for the expression of this gene across all brain regions? And for the brain region with the highest expression? (0.5 point)

**Q1.9**  Does PGBD1 look consistently highly expressed across subjects in this top brain region? For full points, consider patterns of PGBD1 expression across different subjects and brain regions. Use figures and/or statistics to help you decide. (2 points)

**Q1.10**   Looking at a coarser level of analysis, in which (larger) brain structures is PGBD1 most highly expressed? (1 point)

In [ ]:
df_col = pd.read_csv('Columns.csv')
df_exp = pd.read_csv('Expression.csv')

df_col.shape
df_exp.shape

In [ ]:
import pandas as pd
import numpy as np

df_exp_transposed = df_exp.T
df_exp_transposed = df_exp_transposed.reset_index()
df_exp_transposed.columns = ["PGBD1"]
df_exp_transposed = df_exp_transposed.iloc[1:].reset_index(drop=True)

print(df_exp_transposed)
print(df_exp_transposed.shape)

In [ ]:
df_col["PGBD1"] = df_exp_transposed["PGBD1"]
df_col.head(10)

In [ ]:
df_col['PGBD1'] = pd.to_numeric(df_col['PGBD1'], errors='coerce')
print(df_col['PGBD1'].dtype)

In [ ]:
mean_expr = df_col.groupby("structure_name")["PGBD1"].mean().sort_values(ascending=False)
print(mean_expr)


Q1.7 Highest mean expression is in the VIIB region.

In [ ]:
missing_values = df_col['PGBD1'].isnull().sum()
print(f"Number of missing values: {missing_values}")

missing_values_VVIB = df_col[df_col['structure_name'] == 'VIIB']['PGBD1'].isnull().sum()
print(f"Number of missing values in VVII region: {missing_values_VVIB}")

Q1.8 The total of missing values for all the regions is 181, for the region with the highest expression (VIIB) 3 values are missing

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

PGBD1_expression = df_col[df_col['top_level_structure_name'] == 'cerebellar cortex']['PGBD1']

subregions = ['I', 'II', 'III', 'IV', 'V', 'VI', 'VIIAf', 'VIIAt', 'VIIB', 'VIIIA', 'VIIIB', 'IX', 'X', 'Crus I', 'Crus II']
cerebellar_subregions_df = df_col[df_col['structure_name'].isin(subregions)]
subregion_stats = cerebellar_subregions_df.groupby('structure_name')['PGBD1'].describe()[['count', 'mean', 'std', 'min', 'max']]
subregion_stats_sorted = subregion_stats.sort_values(by='mean', ascending=False)

print(subregion_stats_sorted)


sns.boxplot(x='structure_name', y='PGBD1', data=cerebellar_subregions_df, showcaps=True, showfliers=False)

plt.xticks(rotation=90)
plt.ylabel('PGBD1 Expression (z-value)')
plt.xlabel('Cerebellar Cortex Subregions')
plt.title('PGBD1 Expression Across Cerebellar Cortex Subregions')
plt.tight_layout()
plt.show()

Q1.9 It seems to look consistenly highly expressed in this top brain region. In every brain region, the z-value is positive and lays above 0.5.

In [ ]:
region_stats = df_col.groupby('top_level_structure_name')['PGBD1'].describe()
region_stats_sorted = region_stats.sort_values(by='mean', ascending=False)
print(region_stats_sorted[['mean', 'std', 'min', 'max', 'count']])

Q1.10 The cerebellar cortex is the average highest expresssed PGBD1 (z-value) brain region with 0.976863 and std of 0.854904 compared to the other top brain regions. However, the cerebellar cortex contains an outlier with a z-value of 7.2349 which makes the average expression higher. Altough the other data points are relatively more close to the average (n = 105). Next follows the cingulate gyrus and insula. In contrast to these regions, the cerebellar cortex contains 105 values which is way more than the other regions which contain 21 and 11 respectively. I think this is not a fair comparisment for the PGBD1 expression patterns.

# Task 2: Mouse phenome database

For this task, you will download and import some files from the Mouse Phenome Database (MPD, https://phenome.jax.org). This is a database containing a large amount of mouse experimental data, where information on several phenotypes/genotypes can be found.

Execute the wget commands to download the following files to the colab session and the pd.read_csv commands to load them into python.

Here is a short description of the file contents.

- measurements.csv (= Description, units, and other metadata attributes for all phenotype measures (traits) in MPD. One row per measure.)
- straininfo.csv (= Strain name, vendor, stock number, web page URL, MPD strain ID, and other attributes for each mouse strain in MPD. One row per strain.)
- strainmeans.csv.gz (= Strain averages, SD, SEM, N, CV for all numeric strain survey phenotype measures in MPD.)
- animaldatapoints.csv.gz (= Individual mouse readings for all strain survey phenotype measures. One row per animal reading.)


**Q2.1** How many different mouse strains are listed in the database? (1pt)

**Q2.2** Several strains are not included in projects (n_proj is `nan`). How many strains were actually used in projects? (1pt)

**Q2.3** Which strain was used in the most projects, give the name and id (1 pt)

**Q2.4** Using the measurements and strainmeans file, look up the experiment with measnum = 29508. Give the description of the experiment and the average results for strain C3HeB/FeJ. (**hint:** You should get two values, one for males and one for females) (1 pts)

**Q2.5** For the experiment above (29508) take the results from BALB/cByJ mice (strainid = 50) in animaldatapoints and perform a two-tailed two-sample t-test for `males > females`. What are the t-value and the p-value? (2 pts)

**Q2.6** Plot a dotplot of the the body weight (measnum = 11201 listed in `measurements`) on the y-axis as a function of strain name on the x axis. Color-code sex. Add sensible title and axis labels. Make the axis legend lisible. (**hint:** strain ID is encoded differently in the tables!!!) (2 pts)

**Q2.7** The project  "Wahlsten1" (projsym = "Wahlsten1" listed in `measurements`) examines anxiety behaviour using two assays (open field test) and elevated plus maze (epm) in 20 strains, and across both sexes. Plot the mean values across strain and sex for measnum =  10831 (percentage of time spent 10 cm from the wall, higher values indicate more anxiety) on the x-axis, and measnum = 10852 (percentage of time spent in open arms vs closed. lower values indicate less anxiety) on the y-axis. Add a regression line to the plot. Add sensible title and axis labels. Make the axis legend lisible. Get a correlation value for the two measurments. (3 pts)


In [1]:
# code to download the data

! wget --no-check-certificate https://phenomedoc.jax.org/MPD_downloads/measurements.csv
! wget --no-check-certificate https://phenomedoc.jax.org/MPD_downloads/straininfo.csv
! wget --no-check-certificate https://phenomedoc.jax.org/MPD_downloads/strainmeans.csv.gz
! wget --no-check-certificate https://phenomedoc.jax.org/MPD_downloads/animaldatapoints.csv.gz

--2025-12-11 16:29:36--  https://phenomedoc.jax.org/MPD_downloads/measurements.csv
Resolving phenomedoc.jax.org (phenomedoc.jax.org)... 34.49.178.191
Connecting to phenomedoc.jax.org (phenomedoc.jax.org)|34.49.178.191|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1545658 (1.5M) [text/csv]
Saving to: ‘measurements.csv.5’

measurements.csv.5  100%[===================>]   1.47M  --.-KB/s    in 0.03s   

2025-12-11 16:29:36 (44.1 MB/s) - ‘measurements.csv.5’ saved [1545658/1545658]

--2025-12-11 16:29:36--  https://phenomedoc.jax.org/MPD_downloads/straininfo.csv
Resolving phenomedoc.jax.org (phenomedoc.jax.org)... 34.49.178.191
Connecting to phenomedoc.jax.org (phenomedoc.jax.org)|34.49.178.191|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 366018 (357K) [text/csv]
Saving to: ‘straininfo.csv.5’

straininfo.csv.5    100%[===================>] 357.44K  --.-KB/s    in 0.009s  

2025-12-11 16:29:36 (39.6 MB/s) - ‘straininfo.csv.5’ saved [3

In [2]:
# code to load the data into pandas dataframes
import pandas as pd
measurements = pd.read_csv('measurements.csv')
straininfo = pd.read_csv('straininfo.csv')
strainmeans = pd.read_csv('strainmeans.csv.gz')
animaldatapoints = pd.read_csv('animaldatapoints.csv.gz')

/tmp/ipython-input-3586439961.py:6: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  animaldatapoints = pd.read_csv('animaldatapoints.csv.gz')


In [3]:
straininfo.head()

strainname vendor stocknum panel  mpd_strainid straintype  \
0  BTBR T<+> Itpr3<tf>/J      J   002282   NaN             1         IN   
1               CAST/EiJ      J   000928   NaN             2        IWD   
2            129S1/SvImJ      J   002448   NaN             3         IN   
3                    A/J      J   000646   NaN             4         IN   
4                BALB/cJ      J   000651   NaN             5         IN   

   n_proj  n_snp_datasets mpd_shortname       mginum  \
0   115.0             5.0           NaN  MGI:2162761   
1   152.0             7.0           NaN  MGI:2159793   
2   200.0             7.0           NaN  MGI:3037980   
3   249.0             9.0           NaN  MGI:2159747   
4   111.0             4.0           NaN  MGI:2159737   

                                 url  
0  https://www.jax.org/strain/002282  
1  https://www.jax.org/strain/000928  
2  https://www.jax.org/strain/002448  
3  https://www.jax.org/strain/000646  
4  https://www.jax.org/strain/000651

In [4]:
straininfo.head()
total_strains = straininfo['strainname'].nunique()
print("Q2.1: Total number of strains:", total_strains)

Q2.1: Total number of strains: 4705


In [5]:
strain_projects = straininfo.dropna(subset=['n_proj'])
print("Q2.2: Number of strains used in projects:", len(strain_projects))

Q2.2: Number of strains used in projects: 4366


In [6]:
most_strains = strain_projects.loc[strain_projects['n_proj'].idxmax()]
print(most_strains)

strainname                                 C57BL/6J
vendor                                            J
stocknum                                     000664
panel                                           NaN
mpd_strainid                                      7
straintype                                       IN
n_proj                                        327.0
n_snp_datasets                                 10.0
mpd_shortname                                   NaN
mginum                                  MGI:3028467
url               https://www.jax.org/strain/000664
Name: 6, dtype: object


Q2.3: The C57BL/6J strain has the most projects with 327.

In [7]:
measurement_info = measurements[measurements['measnum'] == 29508]
print(measurement_info)


      measnum mpdsector     projsym            varname  \
2478    29508       css  Eumorphia6  ejection_fraction   

                          descrip units            method intervention  \
2478  cardiac ejection fraction,      %  echocardiography          NaN   

     paneldesc datatype sextested  nstrainstested ageweeks  
2478    inbred      NaN      both             4.0    12wks  


In [8]:
results_C3HeB = strainmeans[(strainmeans['strain'] == 'C3HeB/FeJ') & (strainmeans['measnum'] == 29508)]
print(results_C3HeB)

       measnum            varname     strain  strainid sex  mean  nmice    sd  \
63143    29508  ejection_fraction  C3HeB/FeJ        59   f  63.7    5.0  7.48   
63147    29508  ejection_fraction  C3HeB/FeJ        59   m  59.9    6.0  6.45   

        sem     cv  minval  maxval  zscore  
63143  3.35  0.118    56.1    75.7   -1.40  
63147  2.63  0.108    53.2    70.5   -1.16  


Q2.4: The research measures the percentage of cardiac ejection fractions with echocardiography in both male and female C3HeB/FeJ mice. The female mice have an average of 63.7 (SD: 7.48) and male 59.9 (SD: 6.45). So, the female mice have a higher percentage of cardiac ejections compared to male

In [9]:
animaldatapoints.head()

measnum  strainid sex animal_id  measval  zscore  animal_projid  sample_id
0     1401         4   f       191     73.0   -0.59             14        NaN
1     1401         4   f       192     73.0   -0.59             14        NaN
2     1401         4   f       193     73.0   -0.59             14        NaN
3     1401         4   f       194     73.0   -0.59             14        NaN
4     1401         4   f       195     73.0   -0.59             14        NaN

In [10]:
results_BALB = animaldatapoints[(animaldatapoints['strainid'] == 50) & (animaldatapoints['measnum'] == 29508)]
print(results_BALB)

        measnum  strainid sex animal_id  measval  zscore  animal_projid  \
539291    29508        50   f     35584     76.9    0.66            295   
539292    29508        50   f     35585     73.1    0.11            295   
539293    29508        50   f     35586     66.1   -0.90            295   
539294    29508        50   f     35587     68.8   -0.51            295   
539295    29508        50   f     35588     83.4    1.60            295   
539296    29508        50   f     35589     65.7   -0.96            295   
539314    29508        50   m     35578     68.1   -0.57            295   
539315    29508        50   m     35579     68.2   -0.56            295   
539316    29508        50   m     35580     77.5    0.91            295   
539317    29508        50   m     35581     80.8    1.42            295   
539318    29508        50   m     35582     71.9    0.03            295   
539319    29508        50   m     35583     63.9   -1.23            295   

        sample_id  
5392

In [11]:
from scipy.stats import ttest_ind
female = results_BALB[results_BALB['sex'] == 'f']['measval']
male = results_BALB[results_BALB['sex'] == 'm']['measval']


t_stat, p_val = ttest_ind(male, female, equal_var=False)
print(f"T-statistiek: {t_stat}")
print(f"P-waarde: {p_val}")

T-statistiek: -0.15635276846115717
P-waarde: 0.8788875475092771


Q2.6 Plot a dotplot of the the body weight (measnum = 11201 listed in measurements) on the y-axis as a function of strain name on the x axis. Color-code sex. Add sensible title and axis labels. Make the axis legend lisible. (hint: strain ID is encoded differently in the tables!!!) (2 pts)

In [12]:
measurements.head()

measnum mpdsector   projsym       varname  \
0     1401       css  Johnson1       age_yng   
1     1402       css  Johnson1     click_yng   
2     1403       css  Johnson1   pip8kHz_yng   
3     1404       css  Johnson1  pip16kHz_yng   
4     1405       css  Johnson1  pip32kHz_yng   

                                             descrip units  \
0           age of younger group (local parameter),      d   
1  auditory brainstem response (ABR) threshold, 3...    dB   
2  auditory brainstem response (ABR) threshold, 3...    dB   
3  auditory brainstem response (ABR) threshold, 3...    dB   
4  auditory brainstem response (ABR) threshold, 3...    dB   

               method intervention paneldesc datatype sextested  \
0  colony observation          NaN    inbred      NaN      both   
1            ABR test          NaN    inbred      NaN      both   
2            ABR test          NaN    inbred      NaN      both   
3            ABR test          NaN    inbred      NaN      both   
4            ABR test          NaN    inbred      NaN      both   

   nstrainstested ageweeks  
0            42.0      NaN  
1            42.0  3-12wks  
2            42.0  3-12wks  
3            42.0  3-12wks  
4            42.0  3-12wks